### PR TITLE
Only add definitions for used registers

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch makes the BHive to Exegesis conversion script only include definitions for registers needed by the snippet rather than a random set. This makes the outputs cleaner and will help keep things clean later on when setting a loop register (and also make information available to determine if/where there are conflicts).

### CODE REVIEW

**Comment by mtrofin** [comment]
should it have a test?

**Comment by boomanaiden154** [comment]
Existing tests have been updated (which should be enough coverage here) now that testing infrastructure is  in place.

**Comment by boomanaiden154** [comment]
I accidentally merged in #52 into this branch (tried to do stacked PRs and forgot how the DAG was setup while being overly zealous with the merge button). I can probably fix this if it's really necessary but the only changes from that PR (which was approved) are `json.test`, and lines 305-314.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:235]: With this, we enumerate inputs as well as outputs. Ideally, we should go from `instr_desc.getNumDefs()` to `mc_inst.getNumOperands()` to capture only the inputs.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:244]: This stops at the _last_ super-register. Is there a guarantee that it is also the largest super-register? If not, we should either recurse (whenever we find a proper super-register, step up and look for its own super-registers), or filter by register class and take only the 64-bit superreg, if any.

Though the general idea is probably safer and more general.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:246]: We also need to look at implicit input registers, via `MCInstrDesc::implicit_uses()`.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:248]: Thinking about it a bit more, this should go to a library function where we can also add proper unit tests.

**Review by ondrasej — CHANGES_REQUESTED** [review]

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:244]: From everything that I can tell, yes. It's not explicitly enumerated in the documentation for the iterator, but the iterator internally starts at the given register and then goes up one level at a time (from what I can tell). Would probably be good to open a patch on the LLVM side to document this.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:248]: Good point. Split into #71, where the actual functionality is implemented.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:235]: Good point. Switched over to iterating through the inputs of the register only (and keeping track of what registers get defined in the snippet so we can omit them from the usage list).

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (boomanaiden154) [reply gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:246]: Good point. Functionality for that has been added in #71.

**Comment by boomanaiden154** [comment]
This is now part of a PR stack, depending on #71.

**Comment by ondrasej** [comment]
> This is now part of a PR stack, depending on #71.

SG, let's get #71 merged first.

**Comment by ondrasej** [comment]
I'll wait with reviewing until #71 is merged and we have this based just on main.

**Comment by boomanaiden154** [comment]
> I'll wait with reviewing until https://github.com/google/gematria/pull/71 is merged and we have this based just on main.

I've merge main into the branch, so the diff should be clean now.

File: gematria/datasets/basic_block_utils.cc
Comment (ondrasej) [inline gematria/datasets/basic_block_utils.cc:35]: No action needed, just discussion.

If I understand the comment in [MCRegisterInfo.h](https://github.com/llvm/llvm-project/blob/00c622e596f918d9d83674b58097c8982ae1af95/llvm/include/llvm/MC/MCRegisterInfo.h#L109) correctly, the superregs are stored in a zero-terminated array, so what we see here is probably the terminator (and we need to do something like back() - 1)?

Either way, this seems to be safe, and unless it is documented somewhere that the last value is the largest one, I'd stick with the slower but safe code.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:241]: Very nit: Add a period at the end of the line.

File: gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc
Comment (ondrasej) [inline gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc:237]: The closing apostrophe is missing (we should have `"'\n"` rather than `"\n"` at the end).

**Review by ondrasej — APPROVED** [review]
Looks good. I've added a few minor comments, go ahead with the merge once they are resolved.

File: gematria/datasets/basic_block_utils.cc
Comment (boomanaiden154) [reply gematria/datasets/basic_block_utils.cc:35]: Ah, that would make sense. Sounds good keeping the safe code for now. I guess if it shows up hot in a profile, we can always adjust, but I don't really see that happening.

**Pull request merged**
